### PR TITLE
Enable instance create saga to create a new disk

### LINKED ISSUE #812: Implement disk creation during instance creation
https://github.com/oxidecomputer/omicron/pull/771 adds support for attaching an already created disk during the instance creation saga but leaves out support for creating a disk during instance creation. This issue tracks that work.

### PR DESCRIPTION
**Pull request by zephraph**
Fixes #812

Wires up the instance disk create flow by directly appending the disk create saga as a sub saga. I also corrected the behavior of the attach action by ensuring it doesn't spawn a saga in its action but rather just updates the relevant data in the datastore. There's a TODO to make that functionality more robust, but I covered everything that's handled currently by `instance attach` as defined in nexus... which admittedly isn't much. 

While I was in here I also removed some of the unnecessary behavior around looping through the max number of disks and passing the index down as apart of the saga params. Now the disk create or attach params themselves are passed down and we only loop as many times as is needed. This was an artifact of having static sagas previously. 

There's still more cleanup to do for NICs/external IPs which I haven't tackled and of course more robust tests are required. I'm going to leave the latter up to @smklein as he's been working on that for disk create already.

### CODE REVIEW

File: nexus/src/app/sagas/instance_create.rs [hunk 1]
```diff
@@ -781,11 +796,33 @@ async fn ensure_instance_disk_attach_state(
...
+                    .instance_detach_disk(
+                        &opctx,
+                        &organization_name,
+                        &project_name,
+                        &instance_name,
+                        &disk_name,
+                    )
+                    .await
+            }
+            .map_err(ActionError::action_failed)?;
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:825]: I definitely think this can be cleaned up. The create and attach flows are basically exactly the same.

Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs]: Did the thing.

File: nexus/src/app/sagas/instance_create.rs [hunk 2]
```diff
@@ -705,39 +705,90 @@ async fn sic_allocate_instance_external_ip_undo(
...
+                    &organization_name,
+                    &project_name,
+                    &disk_create.identity.name.clone().into(),
+                )
+                .await
+                .map_err(ActionError::action_failed)?;
+        }
+
+        _ => {}
+    };
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:790]: I'm uncertain if this is sufficient for the implementation or if, like for the NICs, we should handle the case where the disk has already been deleted.

File: nexus/src/app/sagas/instance_create.rs [hunk 3]
```diff
@@ -705,39 +705,90 @@ async fn sic_allocate_instance_external_ip_undo(
...
         return Ok(None);
     }
 
     let disk = &saga_disks[disk_index];
 
+    // TODO-correctness TODO-security It's not correct to re-resolve the
+    // organization and project names now.  See oxidecomputer/omicron#1536.
+    let organization_name: db::model::Name =
+        saga_params.organization_name.clone().into();
+    let project_name: db::model::Name = saga_params.project_name.clone().into();
```
[the remark this answers is not in the document]
Comment (smklein) [reply nexus/src/app/sagas/instance_create.rs:729]: you may want to first deal with the comment below before dealing with this; that should almost certainly be a higher priority

File: nexus/src/app/sagas/instance_create.rs [hunk 4]
```diff
@@ -705,39 +705,90 @@ async fn sic_allocate_instance_external_ip_undo(
...
+            osagactx
+                .nexus()
+                .project_create_disk(
+                    &opctx,
+                    &organization_name,
+                    &project_name,
+                    &create_params,
+                )
+                .await
+                .map_err(ActionError::action_failed)?;
```
Comment (smklein) [inline nexus/src/app/sagas/instance_create.rs:742]: I *think* this may have some idempotency problems as-is -- I also wanna tag in @andrewjstone who actually wrote the subsaga implementation, so please, correct me if I'm wrong.

**TL;DR: I think that you are creating two totally distinct sagas here, rather than actually registering the disk creation as a sub-saga**

AFAICT, this is what you are currently doing in this PR:
- Constructing a [subsaga for each createable disk](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/instance_create.rs#L293-L316), which performs the "create" and "attach" steps as separate saga steps.
- Within the "create" step, we call `osagactx.nexus().project_create_disk(...)`. This validates the creation parameters, but mostly [runs the `DiskCreate` saga](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/disk.rs#L213-L225). I **THINK** this is a slightly unconventional approach to subsagas - rather than actually registering the disk creation steps with the "instance creation" saga, this "create disks" saga node is a *single* node. Yes, it *happens* to call code that launches another saga, but the "instance create" saga has no idea that's happening, so it'll retry the entire node (launching a new saga to create a disk) on failure.

Let's consider the following scenario:
1. We start an instance creation saga, with a request to make a disk named "foobar" as part of the saga
2. We get to this saga node, and launch the disk creation sub-saga to actually create the "foobar" disk.
3. Let's suppose we get most of the way through disk creation, but we fail at some point because Nexus crashes.

On reboot, the following could happen:
- The "disk create" saga could get replayed first. In that case, we would successfully create the disk! However, this "instance create" saga is totally independent, so this node (meaning `sic_create_disks_for_instance`) will be called again. On the second invocation of `project_create_disk`, we'll see that a disk already exists named "foobar", so we would fail. Even worse, we'd unwind the instance creation saga without trying to delete the disk which was created -- it was created in a separate saga!
- The "instance create" saga could get replayed first. In this case, we'd launch an *entirely new* saga by calling `project_create_disk` again, even though we already started doing one before Nexus crashed (so it'll still be attempted in the saga replay). This *might* work, or it *might* fail, depending on how far the other saga got before bailing. Either way, there's a decent chance to leak a disk here, since it'll *also* try to be replayed independently of this saga.

File: nexus/src/app/sagas/instance_create.rs [hunk 5]
```diff
@@ -705,39 +705,90 @@ async fn sic_allocate_instance_external_ip_undo(
 }
 
 /// Create disks during instance creation, and return a list of disk names
-// TODO implement
 async fn sic_create_disks_for_instance(
```
[the remark this answers is not in the document]
Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:708]: Yep, that sounds right to me.

File: nexus/src/app/sagas/instance_create.rs (hunk 4 above)
Comment (smklein) [reply nexus/src/app/sagas/instance_create.rs:742]: I mentioned this in chat, but for posterity:

I think it may be easier to think about this problem if we view it as "construct a single DAG for the instance create saga, that consists of the disk creation" rather than "create a DAG for the instance create saga that *might* spawn totally distinct DAGs for disk creation".

To do this, I'd recommend re-using the [disk_create](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/disk_create.rs#L76-L79) DAG construction method, and using it to create a [`NodeKind::Subsaga`](https://github.com/oxidecomputer/steno/blob/53ee324036d901649718d32acfb4d238fcaf8f20/src/dag.rs#L222) which gets appended somewhere in the instance create DAG

Comment (andrewjstone) [reply nexus/src/app/sagas/instance_create.rs:742]: @smklein is totally correct here. The scenario laid out, and the current implementation here is exactly how subsagas originally worked, which is what made them unsound. A subsaga was just a single node that spawned a new saga. Now, as Sean points out, what steno does is take a `Node` with `NodeKind::Subsaga` and when that node gets appended the Dag that is the subsaga is traversed and the nodes are inserted in the parent saga so that they are run in the parent saga. In essence, every subsaga DAG is merged into the parent DAG. 

Practically speaking, what this means is that you can't use the [subsaga_append](https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/instance_create.rs#L309-L315) call with the [disk_create (`make_saga_dag`)] DAG construction method @smklein suggested because that helper function takes a `DagBuilder` for the child saga and not a `DAG`.

So what you need to do is essentially combine the two methods above, such that you append the subsaga Params to the parent saga as a constant node, and then you append the `Dag` returned from [`make_saga_dag`][https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/disk_create.rs#L76] as a subsaga node. There's an [example](https://github.com/oxidecomputer/steno/blob/main/src/example_provision.rs#L171-L234) in steno that may help you. But it's essentially what happens in `subsaga_append`.

Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:742]: I think there's a similar example in the remove read-only parent saga:
https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/volume_remove_rop.rs#L78-L89
https://github.com/oxidecomputer/omicron/blob/4be42b45142f58799c6a5444370bbdd75c0340de/nexus/src/app/sagas/volume_remove_rop.rs#L113-L129

Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:742]: I paired with @andrewjstone earlier this evening so I think I have a decent grasp of how to resolve this (and clean up some of the linger patterns)

File: nexus/src/app/sagas/instance_create.rs [hunk 6]
```diff
@@ -179,7 +174,7 @@ impl NexusSaga for SagaInstanceCreate {
         // Helper function for appending subsagas to our parent saga.
         fn subsaga_append<S: Serialize>(
             node_basename: &'static str,
-            subsaga_builder: steno::DagBuilder,
+            subsaga_dag: steno::Dag,
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:177]: On @andrewjstone's recommendation I updated this function to take a dag instead of a builder. It's easy enough to call the builder at the append callsite and this lets us reuse the disk create saga's `make_saga_dag` function.

File: nexus/src/app/sagas/instance_create.rs [hunk 7]
```diff
@@ -281,34 +276,51 @@ impl NexusSaga for SagaInstanceCreate {
...
+                            create_params: create_disk.clone(),
+                        },
+                        subsaga_builder,
+                    )?,
+                    &mut builder,
+                    create_disk,
+                    i,
+                )?;
+            }
+        }
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:306]: @smklein this is the bit that wires up the disk create saga as a proper subsaga. We're literally just re-creating the dag and sending it through `subsaga_append`. There might be a better way to handle the `disk_create` params.

File: nexus/src/app/sagas/instance_create.rs [hunk 8]
```diff
@@ -281,34 +276,51 @@ impl NexusSaga for SagaInstanceCreate {
...
-            subsaga_builder.append(Node::action(
-                "create_disk_output",
-                format!("CreateDisksForInstance-{i}").as_str(),
-                CREATE_DISKS_FOR_INSTANCE.as_ref(),
-            ));
             subsaga_builder.append(Node::action(
                 "attach_disk_output",
                 format!("AttachDisksToInstance-{i}").as_str(),
                 ATTACH_DISKS_TO_INSTANCE.as_ref(),
             ));
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:320]: The attach behavior is still wrong. As per the comment below it only needs to operate on the datastore state (for now) which means it'll be a bit closer to the NIC behavior defined elsewhere in this file. I can go through and update that while I'm here. 

https://github.com/oxidecomputer/omicron/blob/09447c20716761f762d00f0cf71dab1d4486e3e7/nexus/src/app/instance.rs#L759-L761

File: nexus/src/app/sagas/instance_create.rs [hunk 9]
```diff
@@ -281,36 +269,51 @@ impl NexusSaga for SagaInstanceCreate {
...
+                            serialized_authn: params.serialized_authn.clone(),
+                            project_id: params.project_id,
+                            create_params: create_disk.clone(),
+                        },
+                        subsaga_builder,
+                    )?,
+                    &mut builder,
+                    create_disk,
+                    i,
+                )?;
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:298]: The thing I'm most concerned about here is the `create_disk` being passed into subsaga_append. I don't know if that's the right params. I suspect what I'm passing into `make_saga_dag` is actually the right params...

File: nexus/src/app/sagas/instance_create.rs [hunk 10]
```diff
@@ -758,64 +723,51 @@ async fn ensure_instance_disk_attach_state(
...
+            .await
+            .map_err(ActionError::action_failed)?;
 
-    let disk = &saga_disks[disk_index];
+    let (.., authz_disk, _db_disk) = LookupPath::new(&opctx, &datastore)
+        .project_id(project_id)
+        .disk_name(&disk_name)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:754]: We really shouldn't be looking up disks by name here, but that's all we have. Technically I could do a special lookup for disks that were previously created and grab their IDs, but for disks that are only being attached all we have is what the API provides. 

This is a bit of an aside to this issue, but I believe that all names should be resolved inside of `http_entrypoints`. Once calls flow further down in the system the should be operating solely on IDs. Whenever we begin to implement RFD-322 we should probably consider that.

File: nexus/src/app/sagas/instance_create.rs [hunk 11]
```diff
@@ -758,64 +735,50 @@ async fn ensure_instance_disk_attach_state(
...
-                        &organization_name,
-                        &project_name,
-                        &instance_name,
-                        &disk_name,
-                    )
-                    .await
-            }
+    if attached {
+        datastore
+            .instance_attach_disk(
```
Comment (andrewjstone) [inline nexus/src/app/sagas/instance_create.rs:769]: Just for my understanding, as I'm not super familiar with this code: Why change the call from using `osagactx.nexus().instance_XXX` to `datastore.instance_XXX`?

It seems you basically duplicated the methods from nexus/src/app/instance.rs

**Review by andrewjstone — COMMENTED** [review]
@zephraph This looks great! It's really nice to see you creating just enough saga nodes necessary to create and attach disks, and not the MAX. I only had one nit and one question. 

This is probably safe to merge, although, I'd like a few others to review. Presumably you tested this all locally and you can now create disks during instance creation :)

File: nexus/src/app/sagas/instance_create.rs [hunk 12]
```diff
@@ -281,36 +276,55 @@ impl NexusSaga for SagaInstanceCreate {
...
-        for i in 0..(MAX_DISKS_PER_INSTANCE as usize) {
-            let disk_params =
-                DiskParams { saga_params: params.clone(), which: i };
+        // Appends the disk create saga as a subsaga directly to the instance create builder.
+        for (i, disk) in params.create_params.disks.iter().enumerate() {
+            if let InstanceDiskAttachment::Create(create_disk) = disk {
+                let subsaga_name =
+                    SagaName::new(&format!("instance-create-disk-{i}"));
+                let subsaga_builder = DagBuilder::new(subsaga_name);
+                let params = &disk_create::Params {
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/src/app/sagas/instance_create.rs:292]: Didn't even know you could do this.

File: nexus/src/app/sagas/instance_create.rs (hunk 11 above)
Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:769]: Yeah, this was a misunderstanding on my part. I'd mistakenly thought that functions off of `osagactx.nexus()` spawned sagas when that's clearly just an implementation detail. I'll change it back to what it was.

Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:769]: Actually a correction to that: `nexus().instance_attach_disk` isn't the best form. It takes names and does lookups which are resolved before calling the datastore method. You're right that we should reuse `instance_attach_disk` in `/nexus/src/app/instance` but ideally we'd have a different form that takes pre-resolved arguments like the datastore method. 

Ideally we want to resolve as many names as possible at the very start of a saga (or better, before the saga is invoked). 

Do folks think I should change it back or is this form acceptable until we're able to follow up with a better interface for attaching? cc @smklein, @davepacheco. 

This sort of plays into [my previous comment](https://github.com/oxidecomputer/omicron/pull/1873/files/3927efedfa8e89c41460ef33697bdbdefb34ad4d#r1007193729).

File: nexus/src/app/sagas/instance_create.rs [hunk 13]
```diff
@@ -281,36 +276,55 @@ impl NexusSaga for SagaInstanceCreate {
...
             )?;
         }
 
-        // See the comment above where we add nodes for creating NICs.  We use
-        // the same pattern here.
-        for i in 0..(MAX_DISKS_PER_INSTANCE as usize) {
-            let disk_params =
-                DiskParams { saga_params: params.clone(), which: i };
+        // Appends the disk create saga as a subsaga directly to the instance create builder.
+        for (i, disk) in params.create_params.disks.iter().enumerate() {
```
Comment (smklein) [inline nexus/src/app/sagas/instance_create.rs:287]: Nice; stoked to see we are now only generating nodes for the portion of the graph we need.

File: nexus/src/app/sagas/instance_create.rs (hunk 11 above)
Comment (smklein) [reply nexus/src/app/sagas/instance_create.rs:769]: At this point I'd personally err on the side of "doing less name lookups" - the boundary between nexus and datastore is really thin from the point-of-view of sagas. The datastore itself should provide necessary encapsulation for safe DB access, but forcing sagas to go through nexus isn't adding much IMO.

**Review by smklein — COMMENTED** [review]
Have you looked at the CI failures? I'm not sure if they're flakes, or related to this PR

File: nexus/src/app/sagas/instance_create.rs (hunk 10 above)
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:765]: 💯 is that not worth doing here?  If not, I wonder if we should keep the comment that was here before (the TODO-security TODO-correctness one that references omicron#1536).

File: nexus/src/app/sagas/instance_create.rs (hunk 11 above)
Comment (davepacheco) [reply nexus/src/app/sagas/instance_create.rs:769]: A few things:

- Yes, I think at some point we discussed having the Nexus-level functions accept `LookupPath` objects rather than "names" so that we could use them from places that had ids or names.
- Yes, I think sagas should use the datastore freely when that's the right level of abstraction.
- At the same time, it might also be appropriate to use the Nexus-level functions when there's real application-level logic that has to happen.  For example, I think it'd be better for sagas that need to update instance state to use the app-level instance_set_runtime().  In this specific case, it seems like the app-level wrapper doesn't do anything _but_ the name resolution and then call the datastore function.  So it seems fine.

**Review by davepacheco — COMMENTED** [review]
This looks good!  Such a huge improvement, not just to what's here but the API and (I expect) the console, too.

I know @smklein is still working on examples for doing unit tests of sagas.  But I think we could still have end-to-end tests that exercise a few key cases here:

- an instance create that creates a disk, attaches another disk, and succeeds
- an instance create that creates one disk, attaches another disk, and fails...we should at least verify after this that the created disk was destroyed and the other disk is detached.  It'd be nice if we can make sure it fails in a way that _did_ create the first disk and/or attach the other disk so that we're exercising the undo actions.

Even if we also had unit tests already I think we'd still want these basic end-to-end tests.

File: nexus/src/app/sagas/instance_create.rs (hunk 10 above)
Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:765]: I can re-add the comment. I still generally believe that we shouldn't just handle this in a one-off way, but more as a holistic solution. I know that's a non-trivial amount of work, but I still believe it to be important.

**Comment by zephraph** [comment]
> Have you looked at the CI failures? I'm not sure if they're flakes, or related to this PR

I've looked at them, yeah. A snapshot integration test is timing out. I'm trying to dig into what's going on there.

File: nexus/src/app/sagas/instance_create.rs [hunk 14]
```diff
@@ -758,64 +735,52 @@ async fn ensure_instance_disk_attach_state(
...
-    let opctx =
-        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
-
-    let saga_disks = &saga_params.create_params.disks;
-    let instance_name =
-        db::model::Name(saga_params.create_params.identity.name);
+    let params = sagactx.saga_params::<DiskAttachParams>()?;
+    let datastore = osagactx.datastore();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
```
Comment (zephraph) [inline nexus/src/app/sagas/instance_create.rs:741]: So @smklein, @andrewjstone this lookup is failing to resolve which is causing the tests to hang. It's not clear to me _why_ this isn't resolved. 

<details>
<summary>Stack trace</summary>

```
thread 'integration_tests::snapshots::test_snapshot' panicked at 'no ancestor called "instance_id"', /Users/just-be/.cargo/registry/src/github.com-1ecc6299db9ec823/steno-0.2.0/src/saga_exec.rs:1914:32
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/panicking.rs:556:5
   1: core::panicking::panic_fmt
             at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/core/src/panicking.rs:142:14
   2: steno::saga_exec::ActionContext<UserType>::lookup::{{closure}}
             at /Users/just-be/.cargo/registry/src/github.com-1ecc6299db9ec823/steno-0.2.0/src/saga_exec.rs:1914:32
   3: core::option::Option<T>::unwrap_or_else
             at /rustc/8b705839cd656d202e920efa8769cbe43a5ee269/library/core/src/option.rs:825:21
   4: steno::saga_exec::ActionContext<UserType>::lookup
             at /Users/just-be/.cargo/registry/src/github.com-1ecc6299db9ec823/steno-0.2.0/src/saga_exec.rs:1911:20
   5: omicron_nexus::app::sagas::instance_create::ensure_instance_disk_attach_state::{{closure}}
             at ./src/app/sagas/instance_create.rs:741:23
```

</details>

`sic_create_instance_record` performs the same lookup. The ordering seems fine... I feel like I'm missing something basic here, I'm just not sure what.

Comment (zephraph) [reply nexus/src/app/sagas/instance_create.rs:741]: @smklein cleared this up for me in chat. So, for some reason in my head I was thinking that this was just a node in the top level instance graph and not a subsaga. It _is_ a subsaga though meaning this lookup wouldn't work because it's in the parent's saga, not the current saga. @smklein pointed out I could just pass `instance_id` down as a param, so I'm going to do that.

**Comment by zephraph** [comment]
> an instance create that creates one disk, attaches another disk, and fails...we should at least verify after this that the created disk was destroyed and the other disk is detached. It'd be nice if we can make sure it fails in a way that did create the first disk and/or attach the other disk so that we're exercising the undo actions.

@davepacheco I think the last test I added tests this case sufficiently. It pre-creates two disks, one of which is faulted. It creates the boot disk, attaches the first good disk, and then tries to attach the faulted disk which results in the saga failing. The test tries to ensure that the same number of disks exist in the expected state post cleanup.

File: nexus/tests/integration_tests/instances.rs [hunk 15]
```diff
@@ -1901,6 +1901,212 @@ async fn test_instance_fails_to_boot_with_disk(
...
+            .body(Some(&instance_params))
+            .expect_status(Some(http::StatusCode::BAD_REQUEST));
+
+    let _response = NexusRequest::new(builder)
+        .authn_as(AuthnMode::PrivilegedUser)
+        .execute()
+        .await
+        .expect("Expected instance creation to fail!");
+
+    // Assert disks are created and attached
```
Comment (smklein) [inline nexus/tests/integration_tests/instances.rs:2089]: In this case, aren't you explicitly testing that they are *not* attached?

File: nexus/tests/integration_tests/instances.rs [hunk 16]
```diff
@@ -1901,6 +1901,212 @@ async fn test_instance_fails_to_boot_with_disk(
...
+    .all_items;
+    assert_eq!(disks.len(), 2);
+
+    for (i, disk) in disks.iter().enumerate() {
+        if i == 0 {
+            assert_eq!(disk.state, DiskState::Detached);
+        } else {
+            assert_eq!(disk.state, DiskState::Faulted);
+        }
+    }
```
Comment (smklein) [inline nexus/tests/integration_tests/instances.rs:2033]: IMO this would be more readable as:

```rust
assert_eq!(disks[0].state, DiskState::Detached);
assert_eq!(disks[1].state, DiskState::Faulted);
```

If you wanted, we could also check UUIDs here, like:

```rust
assert_eq!(disks[0].state, regular_disk.id);
assert_eq!(disks[0].state, DiskState::Detached);

assert_eq!(disks[1].identity.id, faulted_disk.id);
assert_eq!(disks[1].state, DiskState::Faulted);
```

Then we could call this same function *after* the failed instance creation function, to assert that the pre-existing disks appear the same.

File: nexus/tests/integration_tests/instances.rs [hunk 17]
```diff
@@ -1901,6 +1901,212 @@ async fn test_instance_fails_to_boot_with_disk(
...
```
Comment (smklein) [inline nexus/tests/integration_tests/instances.rs:2107]: Seems like it's more important to assert that the disks are the *same* disks, not just that we have two with these particular states. See my comment above re: checking UUIDs; I think that would help.

**Review by smklein — APPROVED** [review]
Looks good - only a couple minor suggestions for the tests!

File: nexus/tests/integration_tests/instances.rs (hunk 15 above)
Comment (zephraph) [reply nexus/tests/integration_tests/instances.rs]: Yep, good catch!

**Review by davepacheco — APPROVED** [review]
It'd be nice to have a comment on the tests saying what they're testing.  Either way, this looks good!
